File: javascript/packages/linter/src/rules/html-require-closing-tags.ts
import { ParserRule } from "../types.js"
import { BaseRuleVisitor } from "./rule-utils.js"

import type { UnboundLintOffense, LintContext, FullRuleConfig } from "../types.js"
import type { HTMLOmittedCloseTagNode, ParseResult, ParserOptions } from "@herb-tools/core"

class RequireClosingTagsVisitor extends BaseRuleVisitor {
  visitHTMLOmittedCloseTagNode(node: HTMLOmittedCloseTagNode): void {
    const tagName = node.tag_name?.value
    if (!tagName) return

    this.addOffense(
      `Missing explicit closing tag for \`<${tagName}>\`. Use \`</${tagName}>\` instead of relying on implicit tag closing.`,
      node.location
    )
  }
}

export class HTMLRequireClosingTagsRule extends ParserRule {
  static autocorrectable = false
  static ruleName = "html-require-closing-tags"
  static introducedIn = this.version("0.9.0")

  get defaultConfig(): FullRuleConfig {
    return {
      enabled: true,
      severity: "error",
    }
  }

  get parserOptions(): Partial<ParserOptions> {
    return { strict: false }
  }

  check(result: ParseResult, context?: Partial<LintContext>): UnboundLintOffense[] {
    const visitor = new RequireClosingTagsVisitor(this.ruleName, context)

    visitor.visit(result.value)

    return visitor.offenses
  }
}

File: rust/herb-linter/src/rules/html_require_closing_tags.rs
use herb::nodes::HTMLOmittedCloseTagNode;
use herb::Visitor;

rule_visitor!(RequireClosingTagsVisitor);
define_parser_rule!(
  HTMLRequireClosingTagsRule,
  "html-require-closing-tags",
  Error,
  RequireClosingTagsVisitor,
  parser_options: { strict: false },
  introduced_in: "0.9.0"
);

impl Visitor for RequireClosingTagsVisitor {
  fn visit_html_omitted_close_tag_node(&mut self, node: &HTMLOmittedCloseTagNode) {
    if let Some(tag_name) = node.tag_name.as_ref().map(|token| token.value.as_str()) {
      self.add_offense(
        format!(
          "Missing explicit closing tag for `<{}>`. Use `</{}>` instead of relying on implicit tag closing.",
          tag_name, tag_name
        ),
        node.location.clone(),
      );
    }
  }
}
